basic example of using assets with io managers (DOC-347) (#23968)
dagster-io/dagster: docs/docs-beta/docs/guides/external-systems/iomanagers.md

sidebar_position: 50
sidebar_label: "Storing data with I/O Managers"
---

> IO Managers in Dagster provide a way to separate the code that's responsible for logical data transformation from the code that's responsible for reading and writing the results.  This can help reduce boiler plate code and make it easy to swap out where your data is stored.

Reviewer: could be worth adding a section that covers swapping out the I/O manager for a different one to cover the "swap out where your data is stored" use case

<details>
<summary>Prerequisites</summary>

- Familiarity with [Assets](/concepts/assets)

> - Familiarity with [Ops and Jobs](/concepts/ops-jobs)

Reviewer: I think the prereq on ops and jobs could be removed

- Familiarity with [Resources](/concepts/resources)
</details>

## Extract IO logic from an asset into a reuseable IO Manager

In many dagster jobs, assets can be broken down as the following steps:
1. reading data from some data store into memory
2. applying some in-memory transform
3. writing the transformed data to some data store

For assets that follow this pattern, using an IO manager can help remove the boiler-plate of the code responsible for reading and writing the data to and from a data source.

For example, both assets in the code below are constructing a Snowflake connection object, reading from an upstream table, applying some in-memory transform, and then uploading the transformed result into a new table in Snowflake.

<CodeExample filePath="guides/external-systems/assets-without-io-managers.py" language="python" title="Assets without IO managers" />

> To switch to using IO managers, we can use the `dagster_snowflake` package's Snowflake / Pandas IO manager instead of the `SnowflakeResource`.  The IO manager reads upstream assets  from the Snowflake table with the shared name.  These are passed into the asset function as Pandas dataframes.  The return type of the asset function should be the transformed Pandas dataframe.  The IO manager is responsible for uploading that dataframe to Snowflake into a table matching the asset definition's name.

Reviewer: ```suggestion
To switch to using IO managers, we can use the Snowflake / Pandas IO manager provided by the `dagster_snowflake_pandas` package. The I/O manager will be used to read and write data instead of the `SnowflakeResource`. 

To load an upstream asset using an I/O manager, the asset is specified as a parameter to the asset function, rather than the `deps` list on the `@asset` decorator. The I/O manager reads the Snowflake table with  the same name as the upstream asset and passes the data into the asset function as a Pandas dataframe.

To store data using an I/O manager, the data should be returned from the asset function. The return type of the function should be updated to be `-> pd.Dataframe`. The I/O manager will then upload the dataframe to a Snowflake table with the same name as the asset.
```

Experimenting with some ways to break this paragraph up a bit. Not super happy with what I've got so far, but I think breaking it into pieces helps with digestibility. Tough to find the balance between how I/O managers work in general and how the Snowflake Pandas I/O manager works specifically....

> <CodeExample filePath="guides/external-systems/assets-with-io-managers.py" language="python" title="Asset with IO managers" />

Reviewer: I think a good resource to link to at the end of the guide could be the list of all I/O managers dagster provides 
https://docs.dagster.io/concepts/io-management/io-managers#built-in-io-managers-1

sidebar_position: 50
sidebar_label: "Storing data with I/O Managers"
---

I/O Managers in Dagster provide a way to separate the code that's responsible for logical data transformation from the code that's responsible for reading and writing the results.  This can help reduce boiler plate code and make it easy to swap out where your data is stored.

<details>
<summary>Prerequisites</summary>

- Familiarity with [Assets](/concepts/assets)
- Familiarity with [Resources](/concepts/resources)
</details>

## Extract I/O logic from an asset into a reuseable I/O Manager

> In many dagster jobs, assets can be broken down as the following steps:

Reviewer 2: ```suggestion
In many Dagster pipelines, assets can be broken down as the following steps:
```

1. reading data from some data store into memory
2. applying some in-memory transform
3. writing the transformed data to some data store

> For assets that follow this pattern, using an I/O manager can help remove the boiler-plate of the code responsible for reading and writing the data to and from a data source.

Reviewer 2: ```suggestion
Using an I/O manager can help simplify the code responsible for reading and writing the data to and from a data source for assets that follow this pattern.
```

For example, both assets in the code below are constructing a DuckDB connection object, reading from an upstream table, applying some in-memory transform, and then writing the transformed result into a new table in DuckDB.

<CodeExample filePath="guides/external-systems/assets-without-io-managers.py" language="python" title="Assets without I/O managers" />

To switch to using I/O managers, we can use the DuckDB / Pandas I/O manager provided by the  `dagster_duckdb_pandas` package.  The I/O manager will be used to read and write data instead of the `DuckDBResource`.

> To load an upstream asset using an I/O manager, that asset is specified as an input parameter to the asset function rather than within the `deps` list on the `@asset` decorator.  The `DuckDBPandasIOManager` specifically reads the DuckDB table with the same name as the upstream asset and passes the data into the asset function as a Pandas dataframe. 

Reviewer 2: ```suggestion
To load an upstream asset using an I/O manager, that asset is specified as an input parameter to the asset function rather than within the `deps` list on the `@asset` decorator.  The `DuckDBPandasIOManager` reads the DuckDB table with the same name as the upstream asset and passes the data into the asset function as a Pandas dataframe. 
```

> To store data using an I/O manager, the data should be returned by the asset function.  The `DuckDBPandasIOManager` specifically takes the returned value from the asset function, which should be a Pandas dataframe, and writes it to the DuckDB file as a table with the same name as the asset.

Reviewer 2: ```suggestion
To store data using an I/O manager, return the data in the asset function. The data returned must be a valid type. The `DuckDBPandasIOManager` accepts Pandas data frames and writes them to DuckDB as a table with the same name as the asset.

Refer to the individual I/O manager documentation for details on valid types and how they store data.
```

<CodeExample filePath="guides/external-systems/assets-with-io-managers.py" language="python" title="Asset with I/O managers" />

## Swapping data stores

> With I/O managers, swapping data stores consists of changing the implementation of the I/O manager resource.  The asset definitions (which should now only contain transformational logic) should not change.

Reviewer 2: ```suggestion
With I/O managers, swapping data stores consists of changing the implementation of the I/O manager resource. The asset definitions, which only contain transformational logic, will not nee to be changed.
```
